File: WORKSPACE/Assignment7/Assignment7.py
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Iris_Data.csv')
X = dataset.iloc[:, :-1].to_numpy()
y = dataset.iloc[:, -1].to_numpy()
mapping = {'Iris-setosa': 0, 'Iris-versicolor': 1, 'Iris-virginica': 2}
y = np.array([mapping[val] for val in y])
#Iris-setosa; Iris-versicolor; Iris-virginica

#no need to split the dataset into a training set and a test set. 


# Feature Scaling
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_scaled = sc.fit_transform(X)





# Using Linear
#from sklearn.svm import SVC
#classifier_linear = SVC(kernel = 'linear')
#classifier_linear.fit(X_scaled, y)
#print("SVC accuracy:", classifier_linear.score(X_scaled, y))


#Using poly
#from sklearn.svm import SVC
#classifier_poly = SVC(kernel='poly', degree=3, C=1.0)
#lassifier_poly.fit(X_scaled, y)
#print("Poly accuracy:", classifier_poly.score(X_scaled,y))


#Using Sigmoid 
#from sklearn.linear_model import LogisticRegression
#classifier_sigmoid = LogisticRegression()
#classifier_sigmoid.fit(X_scaled, y)
#print("Sigmoid accuracy:", classifier_sigmoid.score(X_scaled, y))



#Using rbf
from sklearn.svm import SVC
classifier_rbf = SVC(kernel='rbf')
classifier_rbf.fit(X_scaled, y)
print("rbf accuracy:", classifier_rbf.score(X_scaled,y))



y_pred = classifier_rbf.predict(X_scaled)

#confusion matrix 
from sklearn.metrics import confusion_matrix
print(confusion_matrix(y, y_pred))


from sklearn.decomposition import PCA
from matplotlib.colors import ListedColormap

# Reduce dimensions from 4D to 2D for plotting
pca = PCA(n_components=2)
X_pca = pca.fit_transform(X_scaled)

# Fit the model again using the reduced features
classifier_rbf_2d = SVC(kernel='rbf')
classifier_rbf_2d.fit(X_pca, y)

# Create mesh grid
X1, X2 = np.meshgrid(
    np.arange(start=X_pca[:, 0].min() - 1, stop=X_pca[:, 0].max() + 1, step=0.01),
    np.arange(start=X_pca[:, 1].min() - 1, stop=X_pca[:, 1].max() + 1, step=0.01)
)

# Predict each point on the grid
Z = classifier_rbf_2d.predict(np.array([X1.ravel(), X2.ravel()]).T)
Z = Z.reshape(X1.shape)

# Plot decision boundaries
plt.figure(figsize=(10, 6))
plt.contourf(X1, X2, Z, alpha=0.75, cmap=ListedColormap(['red', 'green', 'blue']))

# Plot actual data points
for i, color, label in zip(range(3), ['red', 'green', 'blue'], ['Setosa', 'Versicolor', 'Virginica']):
    plt.scatter(X_pca[y == i, 0], X_pca[y == i, 1], 
                color=color, label=label, edgecolor='black')

plt.title('SVM with RBF Kernel (Iris Dataset - PCA 2D)')
plt.xlabel('Principal Component 1')
plt.ylabel('Principal Component 2')
plt.legend()
plt.show()

File: WORKSPACE/Assignment7/Assignment7.r


# Importing the dataset
dataset = read.csv('Iris_Data.csv')


# Encode categorical data 
dataset$Species = factor(dataset$Species,
                        levels = c('Iris-setosa', 'Iris-versicolor', 'Iris-virginica'),
                        labels = c(0, 1, 2))
# datase$Species = as.factor()

# Feature Scaling
scaled_cols = scale(dataset[, 1:2])
dataset[, 1:2] = scaled_cols

library(e1071) # install.packages('e1071')
#linear kernel
#classifier = svm(formula = Species ~ .,
               #  data = dataset,
               #  type = 'C-classification', # epds_regression is the default for SVR
                # kernel = 'linear') # radial is the default for SVR
# Predicting the Test set results
#y_pred = predict(classifier, newdata = dataset)
# Showing the Confusion Matrix and Accuracy
#library(caret)
#cm = confusionMatrix(y_pred, dataset$Species)
#print(cm$table)
#print(cm$overall['Accuracy'])


#poly kernel
#classifier = svm(formula = Species ~ .,
                # data = dataset,
                # type = 'C-classification', # epds_regression is the default for SVR
                # kernel = 'poly',
                # degree = 3) # radial is the default for SVR
# Predicting the Test set results
#y_pred = predict(classifier, newdata = dataset)
# Showing the Confusion Matrix and Accuracy
#library(caret)
#cm = confusionMatrix(y_pred, dataset$Species)
#print(cm$table)
#print(cm$overall['Accuracy'])



#sigmoid 
#classifier = svm(formula = Species ~ .,
#  data = dataset,
# type = 'C-classification', # epds_regression is the default for SVR
#  kernel = 'sigmoid') # radial is the default for SVR
# Predicting the Test set results
#y_pred = predict(classifier, newdata = dataset)
# Showing the Confusion Matrix and Accuracy
#library(caret)
#cm = confusionMatrix(y_pred, dataset$Species)
#print(cm$table)
#print(cm$overall['Accuracy'])



#radial basis
classifier = svm(formula = Species ~ .,
                data = dataset,
                type = 'C-classification', # epds_regression is the default for SVR
                kernel = 'radial') # radial is the default for SVR
# Predicting the Test set results
y_pred = predict(classifier, newdata = dataset)
# Showing the Confusion Matrix and Accuracy
library(caret)
cm = confusionMatrix(y_pred, dataset$Species)
print(cm$table)
print(cm$overall['Accuracy'])



# Create a grid over feature space
x1_range <- seq(min(dataset[, 1]) - 1, max(dataset[, 1]) + 1, by = 0.01)
x2_range <- seq(min(dataset[, 2]) - 1, max(dataset[, 2]) + 1, by = 0.01)
grid <- expand.grid(Sepal.Length = x1_range, Sepal.Width = x2_range)

# Predict class for each grid point
grid$Species <- predict(classifier, newdata = grid)

# Convert prediction to factor for ggplot fill
grid$Species <- factor(grid$Species, levels = c(0, 1, 2),
                       labels = c('Setosa', 'Versicolor', 'Virginica'))

# Plot decision boundaries and data points
ggplot() +
  geom_tile(data = grid, aes(x = Sepal.Length, y = Sepal.Width, fill = Species), alpha = 0.8) +
  scale_fill_manual(values = c('#ffcccc', '#ccffcc', '#ccccff')) +
  geom_point(data = dataset, aes(x = Sepal.Length, y = Sepal.Width, color = Species),
             shape = 21, fill = NA, size = 2, stroke = 1) +
  scale_color_manual(values = c('#cc0000', '#009900', '#0000cc')) +
  labs(title = 'SVM Decision Regions (Sepal Features - Scaled)',
       x = 'Scaled Sepal Length', y = 'Scaled Sepal Width') +
  theme_minimal()









#


